tests: enhance thread cooperation

smlng opened the pull request:
PR updates thread_cooperation test output, fixed list of boards with insufficient memory and adapts parameter for boards if possible.

Further it adds a testrunner script for test automation.

smlng commented:
tested on real hardware for boards: `Arduino-mega2560`, `cc2650stk`, and `samr21-xpro` -> works for all.

cgundogan commented:
Please rebase.

smlng commented:
rebased

aabadie commented on tests/thread_cooperation/tests/01-run.py:
```diff
@@ -0,0 +1,22 @@
+def testfunc(child):
+    # 1st 6LoWPAN fragment
+    child.expect(r"\[START\].*\n")
+    child.expect(u"[SUCCESS]")
```
As stated [here](https://github.com/RIOT-OS/RIOT/pull/7906#discussion_r148817267), the `u` is useless with python 3.
Another thing: either use ``child.expect_exact` or escace the `[ ]` in the `[SUCCESS]` string. For me, `make test` fails on native because of this.

aabadie commented on tests/thread_cooperation/tests/01-run.py:
```diff
@@ -0,0 +1,22 @@
+
+def testfunc(child):
+    # 1st 6LoWPAN fragment
+    child.expect(r"\[START\].*\n")
```
The python expect could check more the test application output.

smlng replied:
why? (IMHO) It just makes the script longer and complicated with no additional use, if no `[SUCCESS]` shows up at the end the test failed, parsing any other output doesn't make sense here. It makes sense for some of our timer tests to check if certain jitter/bounds hold, but that doesn't hold here.

aabadie replied:
with #7845 in, this will help to find a problem if there's any. But I also admit that it somehow duplicate the displayed information between C and Python.

aabadie replied:
This line could be changed into:
```python
child.expect(r"\[START\]*.")
```
I don't think the `\n` is needed and the end of line regexp should be `*.` instead of `.*`.

Maybe it could be interesting to check that the test application returns the right factorial result ? (one can compare with the result of:
```python
math.factorial(12)
```
in the pexpect script

smlng replied:
> check that the test application returns the right factorial 

that's already done in the C code of main, could move to test script but then you don't get the result when just running `make term`

smlng replied:
> and the regexp should be *. instead of .*

? but then its no regex anymore, at least not as I know them. IIRC `*` means 0 or more of whatever expression is before, and `.` means _any valid unicode_, hence `.*` means anything or nothing. Or `\d+` means "any number with at least one digit`. But maybe misunderstood regex?

aabadie replied:
I think your understanding is correct, I'm thinking the same, but inverting them makes the test pass for me.

aabadie replied:
> that's already done in the C code of main, could move to test script but then you don't get the result when just running make term

agreed, so let's keep it as it is.

smlng commented:
@aabadie this test always fails for me on macOS, i.e., for master and this PR - so its none of my changes here. It always drops out of the term session after the first receive in main from the first worker thread. Can you confirm, or does it work on Linux?

aabadie commented:
@smlng, with the change I propose [here](https://github.com/RIOT-OS/RIOT/pull/7836/files#r149692530) this test passes otherwise it fails (pexpect hangs)

aabadie commented on tests/thread_cooperation/Makefile:
```diff
@@ -1,15 +1,28 @@
 DISABLE_MODULE += auto_init
 
+ifneq (,$(filter nucleo32-f042,$(BOARD)))
+PROBLEM ?= 3
```
Could be indented with 2 spaces, here and below

smlng commented:
@aabadie I addressed your comments, and changed the tests script a bit: I removed the check for `[START]` because testing with a nucleo-f411 connecting to serial was so slow that this was never printed and hence always failed ☹️

aabadie commented:
> I removed the check for [START] because testing with a nucleo-f411 connecting to serial was so slow that this was never printed and hence always failed

Very strange. I just tested on a nucleo-l073 with `child.expect("\[START\]*.")` readded and it works like a charm.
Are you sure the tty was ready on your system ? Sometimes it takes time (several seconds) before it can be used.

smlng commented:
@aabadie anyhow, it should work on any system and with any board, it somehow does not for me on macOS with nucleo-f411. Waiting for `[START]` shouldn't make the test fail, as long as `[SUCCESS]` comes last.
